{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460b43de-ac7e-4cc1-994c-9a0a147f5026",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from random import randint\n",
    "\n",
    "\n",
    "def read_subselection(transcript, target_length):\n",
    "    \"\"\"\n",
    "    Return a subsection of a transcript of target length\n",
    "    :param transcript: String\n",
    "    :param target_length: int\n",
    "    :return: Read\n",
    "    \"\"\"\n",
    "    if len(transcript) <= target_length:\n",
    "        return transcript\n",
    "    pos = randint(0, len(transcript) - target_length)\n",
    "    return transcript[pos:pos + target_length]\n",
    "\n",
    "\n",
    "def random_impute_n(read, probability):\n",
    "    \"\"\"\n",
    "    Randomly replace Nuclotides in a string with \"N\"s with a probability\n",
    "    :param read: Nucleotide read\n",
    "    :param probability: Probability a given nucleotide is replaces with N\n",
    "    :return: Modified String\n",
    "    \"\"\"\n",
    "    length = len(read)\n",
    "    my_arr = np.random.choice([0, 1], size=(length,), p=[1 - probability, probability])\n",
    "    return \"\".join([\"N\" if my_arr[i] == 1 else read[i] for i in range(length)])\n",
    "\n",
    "\n",
    "def random_mutation(read, probability):\n",
    "    \"\"\"\n",
    "    Randomly replace Nuclotides in a string with a random mutation with a probability\n",
    "    :param read: Nucleotide read\n",
    "    :param probability: Probability a given nucleotide is replaced with a random nucleotide\n",
    "    :return: Modified String\n",
    "    \"\"\"\n",
    "    length = len(read)\n",
    "    my_arr = np.random.choice([0, 1], size=(length,), p=[1 - probability, probability])\n",
    "    return \"\".join([np.take([\"A\", \"C\", \"G\", \"T\"], randint(0, 3)) if my_arr[i] == 1 else read[i] for i in range(length)])\n",
    "\n",
    "\n",
    "def random_deletion(read, probability):\n",
    "    \"\"\"\n",
    "    Randomly replace deletes Nuclotides in a string with a probability\n",
    "    :param read: Nucleotide read\n",
    "    :param probability: Probability a given nucleotide is removed\n",
    "    :return: Modified String\n",
    "    \"\"\"\n",
    "    length = len(read)\n",
    "    my_arr = np.random.choice([0, 1], size=(length,), p=[1 - probability, probability])\n",
    "    return \"\".join([\"\" if my_arr[i] == 1 else read[i] for i in range(length)])\n",
    "\n",
    "\n",
    "def random_insert(nucl):\n",
    "    \"\"\"\n",
    "    Randomly prepend a given nucleotide with a random nucleotide. EX: \"A\" -> \"CA\"\n",
    "    :param nucl: Nucleotide\n",
    "    :return: 2 nucleotide string\n",
    "    \"\"\"\n",
    "    return np.take([\"A\" + nucl, \"C\" + nucl, \"G\" + nucl, \"T\" + nucl], randint(0, 3))\n",
    "\n",
    "\n",
    "def random_insertion(read, probability):\n",
    "    \"\"\"\n",
    "    Randomly inserts (before) Nuclotides in a string with a random mutation with a probability\n",
    "    :param read: Nucleotide read\n",
    "    :param probability: Probability a given nucleotide is prepended random nucleotide\n",
    "    :return: Modified String\n",
    "    \"\"\"\n",
    "    length = len(read)\n",
    "    my_arr = np.random.choice([0, 1], size=(length,), p=[1 - probability, probability])\n",
    "    return \"\".join([random_insert(read[i]) if my_arr[i] == 1 else read[i] for i in range(length)])\n",
    "\n",
    "\n",
    "def simulate_read(transcript, n_prob, mutation_prob, delete_prob, insert_prob, target_length):\n",
    "    \"\"\"\n",
    "    Takes a transcript, and simulates random possible read from the transcript.\n",
    "    Will replace with N's, mutate, delete or insert with associated probabilities.\n",
    "    :param transcript: Nucleotide transcript\n",
    "    :param n_prob: Probability of replacing a nucleotide with \"N\"\n",
    "    :param mutation_prob: Probability of mutating a nucleotide\n",
    "    :param delete_prob: Probability of deleting a nucleotide\n",
    "    :param insert_prob: Probability of inserting a nucleotide.\n",
    "    :param target_length: Length of read.\n",
    "    :return: Read with (possibly) modified sequence.\n",
    "    \"\"\"\n",
    "    read = random_mutation(read=transcript, probability=mutation_prob)\n",
    "    read = random_insertion(read=read, probability=insert_prob)\n",
    "    read = random_deletion(read=read, probability=delete_prob)\n",
    "    read = random_impute_n(read=read, probability=n_prob)\n",
    "    return read_subselection(transcript=read, target_length=target_length)\n",
    "\n",
    "\n",
    "def shuffle_x_y_together(a, b):\n",
    "    \"\"\"\n",
    "    Shuffles 2 numpy arrays together at the 1st dimension. Keeps them together, such that (Xi,Yi) => (Xj,Xj)\n",
    "    :param a: Numpy Array\n",
    "    :param b: Numpy Array\n",
    "    :return: Shuffled A, Shuffled B\n",
    "    \"\"\"\n",
    "    a_s = a.shape[0]\n",
    "    b_s = b.shape[0]\n",
    "    assert a_s == b_s, \"Arrays must be of same size\"\n",
    "    i = np.argsort(np.random.permutation(a_s))\n",
    "    return a[i, :, :], b[i, :]\n",
    "\n",
    "\n",
    "def build_train_sets(data_samplers,\n",
    "                     sample_size=1000,\n",
    "                     test_size=.2):\n",
    "    \"\"\"\n",
    "    Builds creates a train/test tensor for a round of training. \n",
    "    :param data_samplers: list sfseventools.FastmSampler\n",
    "    :param sample_size: Samples per tensor, default 1000\n",
    "    :param test_size: Test size as percentage of sample size. Default .2\n",
    "    :return: x_train, y_train, x_test, y_test\n",
    "    \"\"\"\n",
    "\n",
    "    length = len(data_samplers)\n",
    "    test_length = int(length * test_size)\n",
    "\n",
    "    x_train = np.concatenate([data_samplers[i].build_tensor(sample_size) for i in range(length)])\n",
    "    y_train = np.concatenate([data_samplers[i].build_output_tensor(sample_size) for i in range(length)])\n",
    "    x_test = np.concatenate([data_samplers[i].build_tensor(sample_size) for i in range(test_length)])\n",
    "    y_test = np.concatenate([data_samplers[i].build_output_tensor(sample_size) for i in range(test_length)])\n",
    "\n",
    "    return x_train, y_train, x_test, y_test\n",
    "\n",
    "\n",
    "def sample_from_transcripts(transcripts, n_prob, mutation_prob, delete_prob, insert_prob, target_length, num_samples):\n",
    "    \"\"\"\n",
    "    Takes a list of nucleotide transcripts, and simulates sets of random possible reads, sampled from the transcripts.\n",
    "    Will replace with N's, mutate, delete or insert with associated probabilities.\n",
    "\n",
    "    :param transcripts: Nucleotide transcripts\n",
    "    :param n_prob: Probability of replacing a nucleotide with \"N\"\n",
    "    :param mutation_prob: Probability of mutating a nucleotide\n",
    "    :param delete_prob: Probability of deleting a nucleotide\n",
    "    :param insert_prob: Probability of inserting a nucleotide.\n",
    "    :param target_length: Length of read.\n",
    "    :param num_samples:\n",
    "    :return: Read with (possibly) modified sequence.\n",
    "    \"\"\"\n",
    "    picks = np.random.choice(len(transcripts), num_samples)\n",
    "    return [simulate_read(transcripts[pick], n_prob, mutation_prob, delete_prob, insert_prob, target_length) for pick in\n",
    "            picks]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69849c6d-eb06-4158-922d-f50a654b2a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fastm_file(input_filepath, uid=None, sep=None, ender=None):\n",
    "    \"\"\"\n",
    "    Ingests a .fastm file, and returns a list of trsancript UIDs and Nucleotide transcripts.\n",
    "\n",
    "    :param input_filepath: Fastm filepath\n",
    "    :param uid: Extraneous chars for read UIDs\n",
    "    :param sep: UID / read separator. Default ')'\n",
    "    :param ender: File endline character. Default '}'\n",
    "    :return: (tuple) (List of Unique Identifiers, List of Transcripts)\n",
    "    \"\"\"\n",
    "    if uid is None:\n",
    "        uid = UID\n",
    "    if sep is None:\n",
    "        sep = SEP\n",
    "    if ender is None:\n",
    "        ender = ENDER\n",
    "    transcript_ids = []\n",
    "    transcripts = []\n",
    "    with open(input_filepath, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            line = line.replace('\\n', '').replace(ender, '')\n",
    "            if uid is not None:\n",
    "                line = line.replace(uid, '')\n",
    "            name, sequence = line.split(sep)\n",
    "            transcript_ids.append(name)\n",
    "            transcripts.append(sequence)\n",
    "    return transcript_ids, transcripts\n",
    "\n",
    "def reads_to_numpy(reads, array_length, randomize_positions, possible_reverse):\n",
    "    \"\"\"\n",
    "    Converts a nucleotide read to a 1-hot numpy array.\n",
    "    :param reads: list of Nucleotide reads\n",
    "    :param array_length: int Array length, must be large enough to fit the provided read \n",
    "    :param randomize_positions: Can the read pe in any position in the read? Else it will at the beginning of the array \n",
    "    :param possible_reverse: Bool Can the read be reversed?\n",
    "    :return: 3-D numpy 1-hot list\n",
    "    \"\"\"\n",
    "    return np.asarray([sequence_to_array(read, array_length, randomize_positions, possible_reverse) for read in reads])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6951d9fa-db1a-4ea1-9051-ac126d8d66be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import uniform\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "class FastmSampler(object):\n",
    "    def __init__(self,\n",
    "                 fastm_file_path,\n",
    "                 n_prob_upper_limit,\n",
    "                 mutation_prob_upper_limit,\n",
    "                 delete_prob_upper_limit,\n",
    "                 insert_prob_upper_limit,\n",
    "                 min_read_length,\n",
    "                 tensor_length,\n",
    "                 b_randomize_location,\n",
    "                 b_randomize_direction,\n",
    "                 training_output,\n",
    "                 uid=None,\n",
    "                 sep=None,\n",
    "                 ender=None):\n",
    "        \"\"\"\n",
    "        Samples a fastm file and builds a training 1-hot tensor. Will impute Ns, mutations etc.\n",
    "        with associated probabities.\n",
    "        \n",
    "        :param fastm_file_path: Path fo fastm file.\n",
    "        :param n_prob_upper_limit: Upper chance to impute N's\n",
    "        :param mutation_prob_upper_limit: Upper chance to impute mutations\n",
    "        :param delete_prob_upper_limit: Upper chance to delete nucleotides\n",
    "        :param insert_prob_upper_limit: Upper chance to insert nucleotides\n",
    "        :param min_read_length: Minimum read length\n",
    "        :param tensor_length: Target tensor length\n",
    "        :param b_randomize_location: Randomize location if read length < tensor length\n",
    "        :param b_randomize_direction: Randomize 3'-5' or 5'-3'?\n",
    "        :param training_output: Truth output vector eg: [1] or [0, 0, 0, 1] etc\n",
    "        :param uid: Extraneous chars for read UIDs (optional)\n",
    "        :param sep: UID / read separator. Default ')' (optional)\n",
    "        :param ender: File endline character. Default '}' (optional)\n",
    "        \"\"\"\n",
    "        self.fastm_file_path = fastm_file_path\n",
    "        self.n_prob_upper_limit = n_prob_upper_limit\n",
    "        self.mutation_prob_upper_limit = mutation_prob_upper_limit\n",
    "        self.delete_prob_upper_limit = delete_prob_upper_limit\n",
    "        self.insert_prob_upper_limit = insert_prob_upper_limit\n",
    "        self.min_read_length = min_read_length\n",
    "        self.tensor_length = tensor_length\n",
    "        self.b_randomize_location = b_randomize_location\n",
    "        self.b_randomize_direction = b_randomize_direction\n",
    "        self.training_output = training_output\n",
    "        ids, self.transcripts = read_fastm_file(self.fastm_file_path, uid=uid, sep=sep, ender=ender)\n",
    "\n",
    "    def build_tensor(self, size):\n",
    "        \"\"\"\n",
    "        Builds a train tensor. Randomly imputes N, mutation etc\n",
    "        :param size: Number of sample to take in training set\n",
    "        :return: Tensor (size x tensor length x 4)\n",
    "        \"\"\"\n",
    "        n_p = uniform(0, self.n_prob_upper_limit)\n",
    "        m_p = uniform(0, self.mutation_prob_upper_limit)\n",
    "        d_p = uniform(0, self.delete_prob_upper_limit)\n",
    "        i_p = uniform(0, self.insert_prob_upper_limit)\n",
    "        target_length = int(round(uniform(self.min_read_length, self.tensor_length), 0))\n",
    "        samples = sample_from_transcripts(self.transcripts,\n",
    "                                          n_prob=n_p,\n",
    "                                          mutation_prob=m_p,\n",
    "                                          delete_prob=d_p,\n",
    "                                          insert_prob=i_p,\n",
    "                                          target_length=target_length,\n",
    "                                          num_samples=size)\n",
    "        return reads_to_numpy(samples, self.tensor_length, self.b_randomize_location, self.b_randomize_direction)\n",
    "\n",
    "    def build_output_tensor(self, size):\n",
    "        \"\"\"\n",
    "        Builds a train tensor. Randomly imputes N, mutation etc\n",
    "        :param size: Number of sample to take in training set\n",
    "        :return: Tensor (size x tensor length x 4)\n",
    "        \"\"\"\n",
    "        return np.asarray([self.training_output for i in range(size)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
